ts/features/bonus/bpd/screens/optInPaymentMethods/OptInPaymentMethodsThankYouDeleteMethodsScreen.tsx

```diff
@@ -0,0 +1,73 @@
+import React, { useEffect } from "react";
+import { useDispatch } from "react-redux";
+import * as pot from "italia-ts-commons/lib/pot";
+import ThankYouSuccessComponent from "../../components/optInPaymentMethods/ThankYouSuccessComponent";
+import { useIOSelector } from "../../../../../store/hooks";
+import { deleteAllPaymentMethodsByFunctionSelector } from "../../../../../store/reducers/wallet/wallets";
+import { isError, isLoading, isUndefined } from "../../model/RemoteValue";
+import {
+  optInPaymentMethodsCompleted,
+  optInPaymentMethodsDeletionChoice
+} from "../../store/actions/optInPaymentMethods";
+import { showToast } from "../../../../../utils/showToast";
+import I18n from "../../../../../i18n";
+import { optInStatusSelector } from "../../store/reducers/details/activation";
+import { LoadingErrorComponent } from "../../../bonusVacanze/components/loadingErrorScreen/LoadingErrorComponent";
+import RetryAfterDeletionFailsComponent from "../../components/optInPaymentMethods/RetryAfterDeletionFailsComponent";
+
+const OptInPaymentMethodsThankYouDeleteMethodsScreen = () => {
+  const dispatch = useDispatch();
+  const deleteAllPaymentMethodsByFunctionStatus = useIOSelector(
+    deleteAllPaymentMethodsByFunctionSelector
+  );
+  const optInStatus = useIOSelector(optInStatusSelector);
+
+  useEffect(() => {
+    // dispatch saga that delete payment methods and update opt-in field
+    dispatch(optInPaymentMethodsDeletionChoice());
+  }, [dispatch]);
+
+  const isOptInStatusLoading = pot.fold(
+    optInStatus,
+    () => true,
+    () => true,
+    _ => true,
+    _ => false,
+    _ => false,
+    _ => true,
+    _ => true,
+    _ => false
+  );
+  // if the payment methods deletion fails show the retry component
+  if (isError(deleteAllPaymentMethodsByFunctionStatus)) {
+    return <RetryAfterDeletionFailsComponent />;
+  }
+
+  // if one between deleteAllPaymentMethodsByFunctionStatus and optInStatus is on loading (or undefined) state show loading component
+  if (
+    isUndefined(deleteAllPaymentMethodsByFunctionStatus) ||
+    isLoading(deleteAllPaymentMethodsByFunctionStatus) ||
+    isOptInStatusLoading
+  ) {
+    return (
+      <LoadingErrorComponent
+        isLoading={true}
+        loadingCaption={I18n.t("global.remoteStates.loading")}
+        onRetry={() => true}
+        testID={"loadingComponent"}
+      />
+    );
+  }
+
+  // if the opt-in choice fails complete the workunit and show an error toast to the user
+  if (pot.isError(optInStatus)) {
+    showToast(I18n.t("bonus.bpd.optInPaymentMethods.thankYouPage.toast"));
+    dispatch(optInPaymentMethodsCompleted());
+    return null;
+  }
```

Comment:
```suggestion
  // perhaps it could be place as first statement in the render block
  if (pot.isError(optInStatus)) {
    return null;
  }
```
Dispatch an action when an async value changes should be done inside an effect.
Dispatch an action inside the render logic could cause some unpredictable behaviors (i.e render could be executed more than one time and it will dispatch the action more than one time)

My proposal is about to put this logic inside an effect that will be executed when this specific value changes
```typescript
useEffect(() => {
    if (pot.isError(optInStatus)) {
      showToast(I18n.t("bonus.bpd.optInPaymentMethods.thankYouPage.toast"));
      dispatch(optInPaymentMethodsCompleted());
    }
  }, [optInStatus, dispatch]);
```
note: I didn't test it

Reply:
Thank you for the explanation. Refactored here [fdfc7d5](https://github.com/pagopa/io-app/pull/3763/commits/fdfc7d5cc4c18174d0d0b40604a72f76fa6cfe31)